python/ray/data/_internal/execution/operators/hash_shuffle.py

```diff
@@ -445,12 +446,143 @@ def __init__(
+        required_cpus = (
+            aggregator_args.get("num_cpus", 1) * self._aggregator_pool.num_aggregators
+        )
+        required_memory = (
+            aggregator_args.get("memory", 0) * self._aggregator_pool.num_aggregators
+        )
+
+        # Check CPU resources
+        total_cpus = cluster_resources.get("CPU", 0)
+        available_cpus = available_resources.get("CPU", 0)
+
+        if required_cpus > total_cpus:
+            raise ValueError(
+                f"Insufficient CPU resources in cluster for hash shuffle operation. "
+                f"Required: {required_cpus} CPUs for {self._aggregator_pool.num_aggregators} aggregators, "
+                f"but cluster only has {total_cpus} total CPUs. "
+                f"Consider reducing the number of partitions or increasing cluster size."
+            )
+
+        if required_cpus > available_cpus:
+            logger.warning(
+                f"Limited available CPU resources for hash shuffle operation. "
+                f"Required: {required_cpus} CPUs, available: {available_cpus} CPUs. "
+                f"Aggregators may take longer to start due to resource contention."
+            )
+
+        # Check memory resources if specified
+        if required_memory > 0:
+            total_memory = cluster_resources.get("memory", 0)
+            available_memory = available_resources.get("memory", 0)
+
+            if required_memory > total_memory:
+                raise ValueError(
+                    f"Insufficient memory resources in cluster for hash shuffle operation. "
+                    f"Required: {required_memory / GiB:.2f} GiB for {self._aggregator_pool.num_aggregators} aggregators, "
+                    f"but cluster only has {total_memory / GiB:.2f} GiB total memory. "
+                    f"Consider reducing the number of partitions or increasing cluster size."
+                )
+
+            if required_memory > available_memory:
+                logger.warning(
+                    f"Limited available memory resources for hash shuffle operation. "
+                    f"Required: {required_memory / GiB:.2f} GiB, available: {available_memory / GiB:.2f} GiB. "
+                    f"Aggregators may take longer to start due to resource contention."
+                )
+
+        logger.debug(
+            f"Resource check passed for hash shuffle operation: "
+            f"required CPUs={required_cpus}, available CPUs={available_cpus}, "
+            f"required memory={required_memory / GiB:.2f} GiB, available memory={available_memory / GiB:.2f} GiB"
+        )
+
+    def _check_aggregator_health(self, timeout_seconds: float = 30.0) -> None:
+        """Check if all aggregators are up and running after a timeout period.
+
+        Args:
+            timeout_seconds: Time to wait before checking aggregator health.
+        """
+        if (
+            self._aggregator_startup_time is None
+            or self._aggregator_monitoring_warned
```

Comment:
We'd warn continuously every 10s (or so) until either all of them start up or user terminates